lint/rule.go

```diff
@@ -14,10 +14,295 @@
 
 package lint
 
+import (
+	"fmt"
+	"strings"
+
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+)
+
 // Rule defines a lint rule that checks Google Protobuf APIs.
 type Rule interface {
-	// Info returns metadata about a rule.
-	Info() RuleInfo
-	// Lint performs the linting process.
-	Lint(Request) ([]Problem, error)
+	// GetName returns the name of the rule.
+	GetName() RuleName
+
+	// GetURI returns the URI where the applicable guideline
+	// is documented. (This should generally be an AIP on https://aip.dev/.)
+	GetURI() string
+
+	// Lint accepts a FileDescriptor and lints it,
+	// returning a slice of Problem objects it finds.
+	Lint(*desc.FileDescriptor) []Problem
+}
+
+// FileRule defines a lint rule that checks a file as a whole.
+type FileRule struct {
+	Name RuleName
```

Comment:
Consider using embedded struct to reduce boilerplate codes, for example,
```golang
type descRule struct {
  Name RuleName
  URI string
}

func (r *descRule) GetName() RuleName {
  return r.name
}

func (r *descRule) GetURI() string {
  return r.uri
}

type FileRule struct {
  descRule
  LintFile func(*desc.FileDescriptor) []Problem
}
// And now you only need to implement Lint function.
```

Reply:
As best as I can tell, an embedded struct _increases_ boilerplate code because every single rule has to know about the level of indirection.

Reply:
I don't think the following codes will increase any boilerplate code for a rule:
```golang
f := FileRule{}
f.Name = "checkFieldName"
f.URI = "http://some-where"
f.LintFile = func(d *desc.FileDescriptor) []Problem {}
```
But you may just want:
```golang
f := FileRule{
Name: "checkFieldName"
URI: "http://some-where"
LintFile: func(d *desc.FileDescriptor) []Problem {}
}
```
which an embedded struct will be an issue.

Reply:
Yes, I want the latter, and embedded structs break it. I would rather have more code in the rule declaration file to hold rule authors harmless.